src/Extensions/Diagnostics.HealthChecks/Internal/ServiceFabricHealthCheckPublisher.cs

```diff
@@ -34,61 +34,99 @@ internal class ServiceFabricHealthCheckPublisher : IHealthCheckPublisher
-
 			if (partition == null)
 			{
-				m_logger.LogWarning(Tag.Create(), "Publisher run before partition provided");
+				m_logger.LogWarning(Tag.Create(), "Publisher run before partition provided.");
 
 				return Task.CompletedTask;
 			}
 
-			partition.ReportPartitionHealth(new SfHealthInformation(SourceId, "Summary", ConvertStatus(report.Status)));
-
-			foreach(KeyValuePair<string, HealthReportEntry> entryPair in report.Entries)
+			// Avoiding repeated pattern matching for each report entry.
+			Action<SfHealthInformation> reportHealth = partition switch
 			{
-				HealthReportEntry entry = entryPair.Value;
-				SfHealthInformation sfHealthInformation = new SfHealthInformation(SourceId, entryPair.Key, ConvertStatus(entry.Status))
-				{
-					Description = CreateDescription(entry)
-				};
+				IStatefulServicePartition statefulPartition => statefulPartition.ReportReplicaHealth,
+				IStatelessServicePartition statelessPartition => statelessPartition.ReportInstanceHealth,
+				_ => (_) => { }, // NOOP, not going to happen, but we need a default case for the switch.
+			};
 
-				try
-				{
-					partition.ReportPartitionHealth(sfHealthInformation);
-				}
-				catch (Exception exception)
+			try
+			{
+				foreach (KeyValuePair<string, HealthReportEntry> entryPair in report.Entries)
 				{
-					m_logger.LogError(Tag.Create(), exception, "Failed to report partition health");
+					cancellationToken.ThrowIfCancellationRequested();
+					reportHealth(BuildSfHealthInformation(entryPair.Key, entryPair.Value));
 				}
+
+				cancellationToken.ThrowIfCancellationRequested();
+				reportHealth(BuildSfHealthInformation(report));
+			}
+			catch (FabricObjectClosedException)
+			{
+				// Ignore, the service instance is closing.
 			}
 
 			return Task.CompletedTask;
 		}
 
-		private string CreateDescription(HealthReportEntry entry)
+		private SfHealthInformation BuildSfHealthInformation(string healthCheckName, HealthReportEntry reportEntry)
 		{
-			// Healthy reports won't be displayed and most of the checks will be healthy,
-			// so we are creating a detailed description only for failed checks and avoiding allocations for healthy
-			if (entry.Status == HealthStatus.Healthy)
+			StringBuilder descriptionBuilder = new StringBuilder();
+			if (!string.IsNullOrWhiteSpace(reportEntry.Description))
 			{
-				return entry.Description;
+				descriptionBuilder.AppendLine($"Description: {reportEntry.Description}");
+			}
+			descriptionBuilder.AppendLine($"Duration: {reportEntry.Duration}.");
+			if (reportEntry.Exception != null)
+			{
+				descriptionBuilder.AppendLine($"Exception: {reportEntry.Exception}");
 			}
 
-			return new StringBuilder()
-				.Append("Exception:").AppendLine(entry.Exception?.ToString())
-				.Append("Description:").AppendLine(entry.Description)
-				.Append("Duration:").AppendLine(entry.Duration.ToString())
-				.Append("Tags:").AppendLine(string.Join(",", entry.Tags))
```

Comment:
Why are we not adding tags anymore? They are not useful at all?

Reply:
No. They are fixed for each health check and usually used to skip execution based on a predicate.